### PR TITLE
`Programming Exercises`: Fix server-side parsing of parametrized test cases from problem statement

### PR DESCRIPTION
**Pull request by ole-ve**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverag
- [X] I implemented the changes with a good performance and prevented too many database calls.
- [X] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
@JohannesStoehr found the issue that parametrized test cases are not parsed correctly from the problem statement. These changes fix this issue.

### Description
Previously, the test case names were split by a comma. This works fine for regular test case (names), but not for parametrized test case (names).

Example of non-parameterized test cases in the problem statement:
`testClass[BubbleSort],testMethods[Context], testMethods[Policy]`

Example of parameterized test cases in the problem statement:
`testInsert(InsertMock, int)[1], testInsert(InsertMock, int)[2], testInsert(InsertMock, int)[3])`

This example shows that just splitting by a comma does not work here. With these changes, the test cases are only split, if every round bracket has been closed. Therefore, a comma between arguments of a parameterized test case is not considered a split between test case names.

### Steps for Testing
Testing these changes on a test server is not possible with additional steps since the task-test mapping on the server-side is not directly shown in the client but is only used for the hint system. Therefore, I suggest either testing this locally and checking the correct mapping in the database or executing the added method with manual examples. However, you can check that the task names are saved correctly:
1. Create a new programming exercise (Java Maven/Gradle)
2. Go to the programming exercise details page.
3. Click the button "Manage hints"
4. Click the blue-colored button to create a new text hint
5. Select the dropdown and verify that all tasks from the problem statement are listed as dropdown values.

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java [hunk 1]
```diff
@@ -133,18 +133,59 @@ private List<ProgrammingExerciseTask> extractTasks(ProgrammingExercise exercise)
...
+            else if (currentChar == ')') {
+                numberUnclosedRoundedBrackets--;
+            }
+
+            currentTestCaseName.append(currentChar);
+
+            // when at the end of the captured names, add the current name to the result
+            if (capturedTestCaseNames.length() == i + 1) {
+                testCaseNames.add(currentTestCaseName.toString());
+            }
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java:186]: Why don't you move this line after the for-loop?

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java [hunk 2]
```diff
@@ -133,18 +133,59 @@ private List<ProgrammingExerciseTask> extractTasks(ProgrammingExercise exercise)
...
+        }
+
+        return testCaseNames.stream().map(String::trim).toList();
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java:189]: I'd already trim the strings in line 168 and the last test case

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java (hunk 1 above)
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java:186]: If a task in the PS does not contain any test case (empty rounded braces), this would otherwise add a test case with an empty name. Of course, you could also check this before the loop.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java (hunk 2 above)
Comment (ole-ve) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java:189]: The idea was to only trim once. But it probably would help the readability.

**Review by akesfeden — APPROVED** [review]
Tested on TS1, works as described.

**Review by JohannesStoehr — APPROVED** [review]
Code looks good 👍

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on ts1, works as fine. All the tasks appear in the dropdown menu.

**Comment by ole-ve** [comment]
@dfuchss
If you face this issue for a Java/Gradle-exercise, you can annotate your JUnit test cases each with @DisplayName("<test-name>") and use <test-name> in the PS for this specific test. For Maven exercises, the annotation seems to have some known issues.

**Comment by dfuchss** [comment]
> @dfuchss If you face this issue for a Java/Gradle-exercise, you can annotate your JUnit test cases each with @DisplayName("<test-name>") and use <test-name> in the PS for this specific test. For Maven exercises, the annotation seems to have some known issues.

For now, we didn't use Parameterized Tests so far :) Therefore, the issue didn't occur for us :) But thanks for the information.

**Review by Hialus — APPROVED** [review]
Code looks good.
A suggestion for improving this further in the future: Allow the escaping of commas with a backslash. So you could also have manual test case names (with `@DisplayName`) that include commas.
